cpp/include/cugraph/graph_mask.hpp

```diff
@@ -0,0 +1,426 @@
+  {
+    if (has_edge_mask_) {
+      // TODO: This is going to implicitly limit edge_t to uint32_t. Need to support 64-bit types as
+      // well
+      return detail::_get_bit<mask_t>(edges_, static_cast<mask_t>(edge_offset));
+    } else {
+      return !complement_;
+    }
+  }
+
+  /**
+   * Add specific vertex to mask
+   * @param vertex id of vertex to mask
+   * @return
+   */
+  __device__ void mask_vertex(vertex_t vertex)
+  {
+    RAFT_EXPECTS(has_vertex_mask_, "Vertex mask needs to be initialized before it can be used");
+    detail::_set_bit<mask_t>(vertices_, static_cast<mask_t>(vertex), true);
+  }
+
+  /**
+   * Add specific edge to mask
+   * @param edge_offset offset of edge to mask
+   * @return
+   */
+  __device__ void mask_edge(edge_t edge_offset)
+  {
+    RAFT_EXPECTS(has_edge_mask_, "Edge mask needs to be initialized before it can be used");
+    detail::_set_bit<mask_t>(edges_, static_cast<mask_t>(edge_offset), true);
+  }
+
+  /**
+   * Are masks complemeneted?
+   *
+   * - !complemented means masks are inclusive (masking in)
+   * - complemented means masks are exclusive (masking out)
+   */
+  __host__ __device__ bool is_complemented() const { return complement_; }
+
+  /**
+   * Has the edge mask been initialized?
+   */
+  __host__ __device__ bool has_edge_mask() const { return has_vertex_mask_; }
+
+  /**
+   * Has the vertex mask been initialized?
+   */
+  __host__ __device__ bool has_vertex_mask() const { return has_edge_mask_; }
+
+  /**
+   * Get the vertex mask
+   */
+  __host__ __device__ mask_t* get_vertex_mask() { return vertices_; }
+
+  /**
+   * Get the edge mask
+   */
+  __host__ __device__ mask_t* get_edge_mask() { return edges_; }
+
+ protected:
+  raft::handle_t const& handle_;
+  bool has_vertex_mask_{false};
+  bool has_edge_mask_{false};
+  vertex_t n_vertices_;
+  edge_t n_edges_;
+  bool complement_{false};
+  mask_t* vertices_;
+  mask_t* edges_;
+};
+
+/**
+ * An owning container object to manage separate bitmasks for
+ * filtering vertices and edges. A compliment setting
+ * determines whether the value of 1 for corresponding
+ * items means they should be masked in (included) or
+ * masked out (excluded).
+ *
+ * Creating this object does not allocate any memory on device.
+ * In order to start using and querying the masks, they will
+ * need to first be initialized.
+ *
+ * @tparam vertex_t
+ * @tparam edge_t
+ * @tparam mask_t
+ */
+template <typename vertex_t, typename edge_t, typename mask_t = std::uint32_t>
+struct graph_mask_t {
+ public:
+  using vertex_type = vertex_t;
+  using edge_type   = edge_t;
+  using mask_type   = mask_t;
+  using size_type   = std::size_t;
+
+  ~graph_mask_t()                       = default;
+  graph_mask_t(graph_mask_t&&) noexcept = default;
+
+  graph_mask_t() = delete;
+
+  explicit graph_mask_t(raft::handle_t const& handle,
+                        vertex_t n_vertices,
+                        edge_t n_edges,
```

Comment:
Same here, 

I am not having a single mask_t array will work for edges in multi-GPU. Each GPU holds multiple edge partitions, and shouldn't we better have one array per edge partition?